dlib/geometry/polygon_abstract.h

```diff
@@ -0,0 +1,99 @@
+// Copyright (C) 2022  Davis E. King (davis@dlib.net), Adrià Arrufat
+// License: Boost Software License   See LICENSE.txt for the full license.
+#undef DLIB_POLYGOn_ABSTRACT_H_
+#ifdef DLIB_POLYGOn_ABSTRACT_H_
+
+#include "rectangle.h"
+#include "vector.h"
+
+namespace dlib
+{
+    class polygon
+    {
+        /*!
+            WHAT THIS OBJECT REPRESENTS
+                This object represents a polygon inside a Cartesian coordinate system.
+                It is just a wrapper class for a std::vector<point> with some added
+                methods.
+
+                Note that the origin of the coordinate system, i.e. (0,0), is located
+                at the upper left corner.  That is, points such as (1,1) or (3,5)
+                represent locations that are below and to the right of the origin.
+        !*/
+
+    public:
+        using size_type = std::vector<point>::size_type;
+
+        polygon(const std::vector<point>& points);
+        /*!
+            ensures
+                - #*this represents a polygon defined by points.
+        !*/
+
+        size_type size() const;
+        /*!
+            ensures
+                - returns the number of points in the polygon.
+        !*/
+
+        point& operator[](const size_type idx);
+        /*!
+            ensures
+                - returns the point of the polygon at index idx.
+        !*/
+
+        const point& operator[](const size_type idx) const;
+        /*!
+            ensures
+                - returns the point of the polygon at index idx.
+        !*/
+
+        const point& at(const size_type idx) const;
+        /*!
+            ensures
+                - returns the point of the polygon at index idx.
+        !*/
+
+        /*!
+            ensures
+                - returns smallest rectangle that contains all points in the polygon.
+        !*/
+        rectangle get_rect() const;
+
+        double area() const;
+        /*!
+            ensures
+                - If you walk the points of #*this in order to make a closed polygon, what is its
+                  area?  This function returns that area.  It uses the shoelace formula to
+                  compute the result and so works for general non-self-intersecting polygons.
+        !*/
+
+        template <typename alloc>
+        void get_convex_shape (
+            const long top,
+            const long bottom,
+            std::vector<double, alloc>& left_boundary,
+            std::vector<double, alloc>& right_boundary
+        ) const;
+        /*!                                                                                
+            requires                                                                       
+                - 0 <= top <= bottom                                                       
+            ensures                                                                        
+                - interprets points as the coordinates defining a convex polygon.  In      
```

Comment:
Add a note that this isn't illegal to call if it's non-convex but it just won't yield a convex shape.  It will always yield an enclosing shape though, I think.

Can you add a few unit tests for this class? :)

Reply:
Sure! I just put this together quickly last night before going to sleep. I submitted the PR once I got it working for my use case.

Reply:
No worries :D